### ISSUE TITLE
Request for 'primitives' topOfPane and endOfPane to go into leo Qt core

### ISSUE THREAD

**Issue by jknGH** [leo-editor/leo-editor#2910]
Hi Edward
    pretty small beer, but:

A long time ago in the Leo Usenet group: eg. [https://groups.google.com/g/leo-editor/c/o8XVdNf5Lrg/m/JzEzOws9AwAJ](https://groups.google.com/g/leo-editor/c/o8XVdNf5Lrg/m/JzEzOws9AwAJ)

I asked about putting a couple of navigation 'primitives' into Leo's Qt core. I resurrected my playing recently, and although to my surprise my code still mostly worked (I had to make some changes due to improved architecture of the lossage mechanism) the primitives aren't 'yet' in the core. I think it would be nice for them to be so.

The primitives are to move the text insertion cursor to the beginning, or end, of the 'pane' - the visible part of a node on screen.
My current implementations are below. I don't claim that these are the best, but they seem to work for me.

```
from leo.core.leoQt import QtCore
def topOfPane():
    """ move the text cursor to the first character visible in the body pane
    """
    w = c.frame.body.widget

    # create a QPoint of the top of the widget (client area)
    top_left = QtCore.QPoint(0, 0)
   
    # get the (text)'position' within the widget of the top left QPoint
    start_pos = w.cursorForPosition(top_left).position()
   
    # create a cursor and set it to this position
    cursor = w.textCursor()
    cursor.setPosition(start_pos)
    # set the Leo cursor from this
    w.setTextCursor(cursor)
   
    c.bodyWantsFocusNow()
```

and

```
from leo.core.leoQt import QtCore

def endOfPane():
    """ move the text cursor to the last character visible in the body pane
    """
    w = c.frame.body.widget
    # get viewport of the body widget
    vp = w.viewport()

    # get the coordinates of the bottom of the visible area
    vWidth = vp.width() -1
    vHeight = vp.height() -1
    # g.es("width", vWidth, "height", vHeight)

    # create a QPoint from this
    bottom_right = QtCore.QPoint(vWidth, vHeight)

    # get the (text)'position' within the widget of the bottom right QPoint
    end_pos = w.cursorForPosition(bottom_right).position()

    # create a cursor and set it to this position
    cursor = w.textCursor()
    cursor.setPosition(end_pos)
    # set the Leo cursor from this
    w.setTextCursor(cursor)

    c.bodyWantsFocusNow()
```

FWIW here is my usage below; these are separately bound to the Numeric Keypad 'Home' and 'End' key. They allow an emulation of the Brief/CRiSP editor use of these keys, which I find very convenient:

- press Home key once: move to beginning of line
- press Home key twice: move to beginning of pane
- press Home key three times: move to beginning of buffer
- press End key once: move to end of line
- press End key twice: move to end of pane
- press End key three times: move to end of buffer

```
def do_crisp_home():
    # we'll just work with the keynum, that shouldn't change
    # have to work if less than three keys in lossage
    last_keynum = g.app.lossage[0].keynum
    lossage_len = len(g.app.lossage)
    if lossage_len < 2 or g.app.lossage[1].keynum != last_keynum:
        # g.es("beginning of line")
        c.editCommands.beginningOfLine(None)
    elif lossage_len < 3 or g.app.lossage[2].keynum != last_keynum:
        # g.es("beginning of pane")
        topOfPane()
    else:
        # g.es("beginning of buffer")
        c.editCommands.beginningOfBuffer(None)


def do_crisp_end():
    # we'll just work with the keynum, that shouldn't change
    # have to make sure we work if there are fewer than three entries in lossage
    last_keynum = g.app.lossage[0].keynum
    lossage_len = len(g.app.lossage)
    if lossage_len < 2 or g.app.lossage[1].keynum != last_keynum:
        # g.es("end of line")
        c.editCommands.endOfLine(None)
    elif lossage_len < 3 or g.app.lossage[2].keynum != last_keynum:
        # g.es("end of pane")
        endOfPane()
    else:
        # g.es("end of buffer")
        c.editCommands.endOfBuffer(None)
```

Thanks a lot for Leo!
Jon N

**Comment by tbpassin** [comment]
You want to make sure that your new cursor does not remove the highlighting of the current line.  Highlighting is a fairly new option that you may not know about yet.  You can read about how to enable it under the Help menu: _Help/Open Help Topics/Help-For-Highlight-Current-Line_.

The highlighting is done in _LeoQTextBrowser.highlightCurrentLine()_.  I don't recall if the call `w.setTextCursor(cursor)`  will also cause the new cursor line to get highlighted, but please make sure this actually happens.

Similarly there is a newish option to draw a right margin guideline.  There is also a Help topic for it - _Help-For-Right-Margin-Guide_.  Again, please make sure that your scripts don't interfere with the right margin guide.

**Comment by jknGH** [comment]
I know about, and use both of the features that you mention and have introduced.

I can say that as written the primitives don't seem to interfere with the highlight current line feature; I haven't checked if they interfere with the right margin feature but can do that sometime soon.

I'm not clear what steps i might need to take before these example implementations are 'accepted'; in the first instance I was hoping Edward could just eyeball them - I am no real Qt programmer! As mentioned, they seemed to be 'acceptable' a few years ago.

I'm happy to be the 'champion' for these and rework if necessary; that might mean me asking some Qt questions though.

**Comment by tbpassin** [comment]
Usually it works like this.  You develop your changes in a separate branch of your own clone of the Leo github repo.  There may be some discussion here or on the Leo Google Group.  

If Edward or someone else doesn't agree that the feature should go into the code base,  it can be hashed out here.  If the issue doesn't get closed at that point, you can go on to submit a Pull Request in Github.  Normally the PR would target the devel branch.

At that time, Edward will probably look at it and possibly tune it up a little if he recognizes something you didn't handle, or could make better use of existing Leo code.  That's usually as much code review as you will get, unless someone else gets interested enough.

Then usually Edward will approve the PR and ask everyone to test the new feature.  Eventually either it will get bug-fixed if needed or make its way into a release branch.

Usually, new features (like the current line highlighting, for example) should be turned off by default, and turned on by a setting that users would set in myLeoSettings.leo.

**Comment by jknGH** [comment]
That all makes sense, and is what I am used to doing in other parts of my life ... but as mentioned, this started out (in the Google Groups link I tried to put up at the top, as: Edward saying, back in 2017:

""
JKN> Edward, I am thinking that it might be worth having the functionality of cursorToTopOfPane() and cursorToEndOfPane() in Leo's (Qt) core?
EKR> ​Yes.  I'll move them there soon.  Thanks for this code.
""
(Edward), I'm not trying to hold you to that... if the workflow Thomas is suggesting is what's needed these days, then fine.

Ah, I see that my link to the original conversation is not working?! Let's try again...

[new cursor motion routines for useful Brief/Crisp-styl Body navigation](https://groups.google.com/g/leo-editor/c/o8XVdNf5Lrg/m/JzEzOws9AwAJ)

**Comment by jknGH** [comment]
Also, it doesn't appear to interfere with the right margin settings - I've just tried on a file that has these enabled

**Comment by tbpassin** [comment]
Oh, good!  Thanks for making sure.

Looking at the linked Groups messages from some years ago, my reaction is that

1. The move to bottom/top of the visible body would be good built-in commands to have;
2. The binding of these commands to multiple keypresses seems pretty specialized to me, and would probably be better activated (by you) for each outline you want to use them with.

I picture a command - maybe "use-brief-keys" - that would monkey-patch c or more likely c.k.  I used this approach when I was prototyping the current-line-highlighting code, and it was very helpful.

Alternatively, if Edward would agree, perhaps there could be a hook that gets called when say the INSERT key gets pressed three times in a row. If not already bound, it would look in the settings for a command to execute. By default, there would be no such command, but for you, your settings file would tell it to run the "use-brief-keys" command.  After that, your binding would be in effect for that outline while it was open.

**Comment by jknGH** [comment]
Erm, I specifically was **not** suggesting the lossage code gets added to core - both in the original thread, and (I thought) in the original issue text above. I thought I was reasonably, clear, but it seems not...

**Comment by tbpassin** [comment]
Probably my reading, not your writing...

**Comment by tbpassin** [comment]
One other thing to check - whether your scripts work for both PyQt5 and PyQt6.  I don't see anything that would be a problem in your code, but a some cases an enum or method in Qt5 gets split out into another package or module for Qt6.  Leo.core.leoQt tries to adjust for this, but it's not always possible.

An example (from the viewrendered3 plugin) is `QWebEngineSettings`, which moved between  _QtWebEngineCore_ and _QtWebEngineWidgets_:
```
   from leo.core.leoQt import QtWebEngineWidgets 
   try:
        from leo.core.leoQt import QtWebEngineCore
        QWebEngineSettings = QtWebEngineCore.QWebEngineSettings
    except Exception:
        QWebEngineSettings = QtWebEngineWidgets.QWebEngineSettings
```

Again, I don't think your code will have a problem, but as always it's best to check.

**Comment by jknGH** [comment]
I don't think I'm in a position to check that.

**Comment by edreamleo** [comment]
@tbpassin @jknGH I plan to make a plugin that adds new commands. That should suffice, imo.

**Comment by edreamleo** [comment]
Closed via PR #2926.
